### ISSUE TITLE
Issues Using on Windows

### ISSUE THREAD

**Issue by andrewM202** [Auburn/FastNoise2#117]
Hello! I have been trying to get FastNoise2 to work on Windows but am new to the C++ build process - I have been able to get it to successfully work on MacOS. Please let me know if my question is better asked on Discord or elsewhere. Currently I have been trying to use VSCode's build tools with FastNoise's precompiled binaries for Clang. The build tools as I have them setup generate this command that runs to build the executable:

`g++ -std=c++17 -fdiagnostics-color=always -g -IC:\Users\hilfu\Desktop\Code\minecraft-opengl/dependencies/include -LC:\Users\hilfu\Desktop\Code\minecraft-opengl/dependencies/library_windows C:\Users\hilfu\Desktop\Code\minecraft-opengl/dependencies/library_windows/FastNoise.lib C:\Users\hilfu\Desktop\Code\minecraft-opengl/*.cpp C:\Users\hilfu\Desktop\Code\minecraft-opengl/glad.c -lglfw3dll -lFastNoise -lFastNoiseD -o C:\Users\hilfu\Desktop\Code\minecraft-opengl/app`

However, this fails with many undefined reference errors, making me think the dlls or .lib files are not linking correctly:

`C:/Users/hilfu/Desktop/Code/minecraft-opengl/dependencies/include/FastNoise/FastNoise.h:34: undefined reference to FastNoise::Add* FastSIMD::New<FastNoise::Add>(FastSIMD::eLevel, void* (*)(unsigned long long, unsigned long long))'`

For reference, attached below is an image of my current file structure. I have also tried using the clang++ compiler instead of g++, but got different errors (which if needed I can try to reproduce again and attach here). If there is any additional information needed to help resolve this, or if there is anything obvious I missed in my inexperience, please let me know. I am hoping I won't have to use FastNoiseLite just for windows!
![image](https://github.com/Auburn/FastNoise2/assets/70407217/d221aa43-8e28-4a03-81bb-7884dc43a10f)

**Comment by Auburn** [comment]
You should use either -lFastNoise or -lFastNoiseD (debug) not both. Your library link folder also looks wrong it should be -LC:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\library_windows

**Comment by andrewM202** [comment]
Thank-you for the input! I have gotten rid of the debug file and fixed the link library folder so it only uses backslashes now, however I seem to still be getting the same undefined reference errors. I also tried forward slashes with no luck. Here is how the command looks now:

`g++ -std=c++17 -fdiagnostics-color=always -g -IC:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\include -LC:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\library_windows C:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\library_windows\FastNoise.lib C:\Users\hilfu\Desktop\Code\minecraft-opengl\*.cpp C:\Users\hilfu\Desktop\Code\minecraft-opengl\glad.c -lglfw3dll -lFastNoise -o C:\Users\hilfu\Desktop\Code\minecraft-opengl\app`

In case it may help I've also dropped here the full error stack as a txt file when running the command. Thank-you for your help.
[full_undefined_error.txt](https://github.com/Auburn/FastNoise2/files/12071698/full_undefined_error.txt)

**Comment by Auburn** [comment]
I assume you are using minGW? I don't think MSVC binaries are compatible with MinGW

**Comment by andrewM202** [comment]
It looks like g++ is mingw, I see that would not work then. I was mostly hoping it would because g++ worked with the MacOS clang binaries. If I switch my project to use ClangCL (with the clang++ command) that should work with the binaries highlighted below then, right? If not I was thinking maybe I should try to port my code over to visual studio since the wiki example uses that and MSVC is used by Visual Studio from what I can tell
![image](https://github.com/Auburn/FastNoise2/assets/70407217/210bf057-a7e7-405d-b5a5-2e2c4ee6c609)

**Comment by Auburn** [comment]
That should work, you will have to use Clang++ outside of MinGW in order to be ABI compatible with the FastNoise2 libs. Alternatively you could compile FastNoise2 libs yourself for Clang on MinGW if you do want to go that route.

GCC MinGW is not supported: https://github.com/Auburn/FastNoise2/issues/69

**Comment by andrewM202** [comment]
Thanks for the suggestion to compile on MinGW! It looks like it worked, I have some other errors in my code but I don't seem to be getting any undefined reference errors anymore, once I get my program fully working shortly and fix these seemingly unrelated errors I'll confirm it definitely worked. 

I actually tried to compile the source before using the wiki, but I only noticed now by default it was compiling for Visual Studio (and thus likely for MSVC)! I used this command so it uses the GCC and G++ compilers/MinGW, and was able to get it to work for the project setup portion:

**`cmake -G "MinGW Makefiles" -S . -B build -D FASTNOISE2_NOISETOOL=OFF -D FASTNOISE2_TESTS=OFF -D BUILD_SHARED_LIBS=ON -D CMAKE_CXX_COMPILER="C:/Users/hilfu/Documents/mingw64/bin/g++.exe" -D CMAKE_C_COMPILER="C:/Users/hilfu/Documents/mingw64/bin/gcc.exe"`**

![image](https://github.com/Auburn/FastNoise2/assets/70407217/6edac2cd-f20b-40a1-aa14-83f44ea85400)


It may be useful to add a note to this wiki page ([linked here](https://github.com/Auburn/FastNoise2/wiki/1:-Compiling-FastNoise2)) that you'll need to add `-G "MinGW MakeFiles` and the extra options for the `CMAKE_CXX_COMPILER` and `CMAKE_C_COMPILER` to compile it for mingw. 

Thanks for your help again!

**Comment by andrewM202** [comment]
Ah, unfortunately I misread and you are right - I got no linker errors compiling with g++ but the program crashes when ran with any FastNoise2 command that actually generates noise. 

I have since tried using clang++ instead, and get no linker errors, but for some reason instead get this error when running program:
![image](https://github.com/Auburn/FastNoise2/assets/70407217/54db5a6b-4bd9-44bd-8363-a0826aa7e3c0)

However, when compiling with clang using the command below I don't even get a libFastNoise.dll so not sure where that is coming from. 
`clang++ -std=c++17 -fdiagnostics-color=always -g -IC:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\include -LC:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\library_windows C:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\library_windows\glfw3dll.lib C:\Users\hilfu\Desktop\Code\minecraft-opengl\dependencies\library_windows\FastNoiseD.lib C:\Users\hilfu\Desktop\Code\minecraft-opengl\*.cpp C:\Users\hilfu\Desktop\Code\minecraft-opengl\glad.c -lglfw3dll -lFastNoiseD -o C:\Users\hilfu\Desktop\Code\minecraft-opengl\app`

![image](https://github.com/Auburn/FastNoise2/assets/70407217/d5c56584-78f8-4faf-87b4-7ecf36b6e61e)

On another note, for some reason when I compile the FastNoise2 source with clang I cannot get the release version only the debug version but I don't think that should matter. I might need to mess around with this a bit more to see if I can get this to work. 
![image](https://github.com/Auburn/FastNoise2/assets/70407217/c1b89db1-8188-40a2-8b86-3270cf9da320)

**Comment by andrewM202** [comment]
Aaahhhh! There we go! For some reason the terrain is see-through (will have to figure that one out lol some other weird windows bug Im sure) but it does look to be generating now with the FastNoise2 noise on windows hah. Thanks so much for all the help Auburn!
![image](https://github.com/Auburn/FastNoise2/assets/70407217/d65189e8-82d5-4e7b-b7ee-6a14b55e6428)

Just for reference, I was able to build FastNoise fully like you suggested in the release mode with this command:
`cmake -G "Ninja" -S . -B build -D FASTNOISE2_NOISETOOL=OFF -D FASTNOISE2_TESTS=OFF -D BUILD_SHARED_LIBS=ON -D CMAKE_CXX_COMPILER="C:/Program Files/LLVM/bin/clang++.exe" -D CMAKE_C_COMPILER="C:/Program Files/LLVM/bin/clang.exe" -D CMAKE_BUILD_TYPE=Release`

I realized that libFastNoise.dll error was because I was building my project with `-o app` which built an app.exe with g++, but which only build an app file with clang++. After fixing that and rebuilding with the new DLLs it looks like it is working!

If you would like I can see at making a pull request with this information on how to build for clang in the wiki - let me know. Thanks for all the help again!
